Beryllium lnwallet 20220414

eoliveros opened the pull request:
- wrap the content of the pages in an alert window.
  - lightning getinfo
  - new address
  - transaction history
  - (all) pages
- withdrawal/psbt:
  - the option to send sat or btc
  - showing the balance in both sat and btc
- sign psbt:
  - shows on-chain balance in btc and sat
- broadcast psbt:
  - shows on-chain balance in btc and sat
- create invoice:
  - checks the amount is not "0"
  - ln-channel balance is now showing
- pay invoice:
    - ln-channel balance is now showing
- invoice history:
  - deleted this page from index
- payment history:
  - deleted this page from index
  - removed route
- open channel:
  - shows on-chain balance in sat
- list forwards:
  - deleted this page from index
- lightning transaction history:
  - combined the invoice and payment history in a series rather than table.
  - shows the ln-channel balance

djpnewton commented on ln.py:
```diff
@@ -116,7 +116,9 @@ def list_invoices(self):
             expires_at = invoice["expires_at"]
-            amount_msat = invoice["amount_msat"]
+            amount_msat = 0
+            if amount_msat:
```
this condition will never be true (```amount_msat = 0```)

eoliveros replied:
reverted back to line 119 when there is no amount_msat = 0 and condition and getting an error:
```
app_1                       |     received_txs = rpc.list_invoices()
app_1                       |   File "/app/ln.py", line 119, in list_invoices
app_1                       |     amount_msat = invoice["amount_msat"]
app_1                       | KeyError: 'amount_msat'
```

will look at it further and see if i can figure this out.

eoliveros replied:
since this is transaction related, changed it.

djpnewton commented on ln.py:
```diff
@@ -128,10 +130,42 @@ def list_invoices(self):
             if status == 'paid':
                 pay_index = invoice["pay_index"]
                 amount_received_msat = invoice["amount_received_msat"]
+                amount_received_sats = _msat_to_sat(invoice["amount_received_msat"])
```
```suggestion
                amount_received_sats = _msat_to_sat(amount_received_msat)
```

eoliveros replied:
changed.

djpnewton commented on ln.py:
```diff
@@ -128,10 +130,42 @@ def list_invoices(self):
+            fees_sats = None
+            if status == "complete":
+                paid_at = created_at
+                paid_date = datetime.datetime.fromtimestamp(paid_at, pytz.timezone('Pacific/Auckland'))
```
we dont want the timezone hardcoded here, we should use the server timezone

eoliveros replied:
removing the pytz.timezone would result on everything being defaulted to UTC.
will most likely need to write a function that would make a os.system to query the timezone being used by the server. need to investigate further.

djpnewton replied:
we can set the timezone in the container if we want to

eoliveros replied:
trying the following in ln.py:
```
if 'SERVER_TIMEZONE' in os.environ:
    server_timezone = os.environ['SERVER_TIMEZONE']
else:
    server_timezone = "UTC"
```

added a line in .env file for SERVER_TIMEZONE

when checking the list_transactions page, its always coming up as UTC timezone.

ive tried adding app.config in app_core as well. its not working though,

eoliveros replied:
changed so this can be configured via env variable

djpnewton commented on ln_wallet_endpoint.py:
```diff
@@ -52,6 +52,7 @@ def list_txs():
     """ Returns template of on-chain txs """
     rpc = LnRpc()
     transactions = rpc.list_txs()
+    logger.info('txs: %s', transactions)
```
remove this?

eoliveros replied:
removed

djpnewton commented on ln_wallet_endpoint.py:
```diff
@@ -121,12 +122,12 @@ def rebalance_channel():
         # I'm assuming there will only be one channel for each node, but I'm
         # using an array in case there's more
         peers[i]["channel_states"] = []
+        can_send_total = 0
```
seems to be unused?

eoliveros replied:
deleted this in my other commit.

djpnewton commented on ln_wallet_endpoint.py:
```diff
@@ -121,12 +122,12 @@ def rebalance_channel():
-                peers[i]["can_send"] += int(channel["msatoshi_to_us"]) / 1000
-                peers[i]["can_receive"] += int(
-                    channel["out_msatoshi_fulfilled"]) / 1000
+                peers[i]["can_send"] += int(channel["spendable_msatoshi"]) / 1000
```
use ```_msat_to_sat``` function

eoliveros replied:
changed to msat_to_sat

djpnewton commented on ln_wallet_endpoint.py:
```diff
@@ -135,7 +136,7 @@ def rebalance_channel():
         peers[i]["sats_total"] = int(peers[i]["sats_total"])
```
these lines seem redundant

eoliveros replied:
removed.

djpnewton commented on ln_wallet_endpoint.py:
```diff
@@ -155,23 +156,27 @@ def pay_invoice():
     received_txs = rpc.list_invoices()
-    return render_template("lightning/received_transactions.html", received_txs=received_txs)
+    send_txs = rpc.list_sendpays()
+    funds_dict=rpc.list_funds()
```
fix spacing around ```=```

eoliveros replied:
fixed

djpnewton commented on ln_wallet_endpoint.py:
```diff
@@ -198,14 +203,17 @@ def channel_opener():
     """ Returns template for creating a PSBT """
     rpc = LnRpc()
     onchain = int(rpc.list_funds()["sats_onchain"]) / 100000000
+    onchain_sats = int(rpc.list_funds()["sats_onchain"])
```
```suggestion
    onchain_sats = int(rpc.list_funds()["sats_onchain"])
    onchain = onchain_sats / 100000000
```

eoliveros replied:
changed.

djpnewton commented on ln_wallet_endpoint.py:
```diff
@@ -235,11 +245,14 @@ def create_psbt():
 def sign_psbt():
+    rpc = LnRpc()
+    onchain = int(rpc.list_funds()["sats_onchain"]) / 100000000
+    onchain_sats = int(rpc.list_funds()["sats_onchain"])
```
```suggestion
    onchain_sats = int(rpc.list_funds()["sats_onchain"])
    onchain = onchain_sats / 100000000
```

eoliveros replied:
changed.

djpnewton commented on ln_wallet_endpoint.py:
```diff
@@ -250,11 +263,14 @@ def sign_psbt():
 def broadcast():
+    rpc = LnRpc()
+    onchain = int(rpc.list_funds()["sats_onchain"]) / 100000000
+    onchain_sats = int(rpc.list_funds()["sats_onchain"])
```
```suggestion
    onchain_sats = int(rpc.list_funds()["sats_onchain"])
    onchain = onchain_sats / 100000000
```

djpnewton commented on templates/lightning/create_psbt.html:
```diff
@@ -35,7 +53,7 @@ <h2>Create PSBT/Withdrawal</h2>
             <input type="text" id="amount" name="amount" class="payment form-control" placeholder="Amount" required>
             <div class="input-group-append">
-                <button type="button" id="max" class="btn btn-sm">Max</button>
+		<button class="btn btn-outline-secondary" type="button" id="max">Max</button>
```
fix spacing

eoliveros replied:
fixed

djpnewton commented on templates/lightning/create_psbt.html:
```diff
@@ -61,7 +80,7 @@ <h2>Create PSBT/Withdrawal</h2>
         bootbox.confirm({
             title: `Confirm Create ${action}`,
-            message: `Total: ${totalBtc} BTC. Do you want to create the ${action}?`,
+		message: `Total: ${totalBtc}${unit}. Do you want to create the ${action}?`,
```
fix spacing

eoliveros replied:
fixed

djpnewton commented on templates/lightning/create_psbt.html:
```diff
@@ -105,7 +124,13 @@ <h2>Create PSBT/Withdrawal</h2>
+            if (unit == 'btc')
+                return ele.val('{{ onchain }}');
+            else if (unit == 'sat')
+		return ele.val('{{ onchain_sats }}');
```
fix spacing

eoliveros replied:
changed the spacing

djpnewton commented on templates/lightning/index.html:
```diff
@@ -1,17 +1,18 @@
         <div class='dashboard'>
         <p>on-chain balance<p>
-        <p><center>{{ funds_dict.sats_onchain }} sats</center></p>
+	<p><center>{{ funds_dict.sats_onchain }} <strong>SAT</strong></center></p>
```
fix spacing

eoliveros replied:
fixed.

djpnewton commented on templates/lightning/index.html:
```diff
@@ -1,17 +1,18 @@
         <div class='dashboard'>
         <p>ln-channel balance<p>
-        <p><center>{{ funds_dict.sats_channels }} sats</center></p>
+	<p><center>{{ funds_dict.sats_channels }} <strong>SAT</strong></center></p>
```
fix spacing

eoliveros replied:
fixed

djpnewton commented on templates/lightning/layout.html:
```diff
@@ -3,4 +3,10 @@
\ No newline at end of file
+{% endblock %}
+
+{% block content %}
```
this layout template should not have a content block

djpnewton replied:
you should put a new block in the main layout template called ```navigation-bar``` or something

eoliveros replied:
this is the result when i add a navbar:
![image](https://user-images.githubusercontent.com/47799652/165648664-b8f75fa9-70b5-4c5e-a426-5690d7d06853.png)

its basically creating one new navbar on top of the main navbar. not good.

eoliveros replied:
ill delete the button from the lightning/layout and put it on individual pages instead

djpnewton commented on templates/lightning/lightning_transactions.html:
```diff
@@ -0,0 +1,41 @@
+            <button class={% if tx.amount_sent_sats is defined %}"btn btn-danger disabled"{% elif tx.amount_received_sats is defined %}"btn btn-success disabled"{% endif %} type="button" data-toggle="collapse" data-target="#collapseExample{{loop.index}}" aria-expanded="false" aria-controls="collapseExample">
+                {{tx.paid_date}}
+            </button>
+	    <b>{% if tx.amount_sent_sats is defined %}-{{tx.amount_sent_sats}}{% elif tx.amount_received_sats is defined %}+{{tx.amount_received_sats}}{% endif %} SAT{% if tx.amount_sent_sats is defined%}(fees included){% endif %}</b>
```
fix spacing

eoliveros replied:
fixed

djpnewton commented on templates/lightning/lightning_transactions.html:
```diff
@@ -0,0 +1,41 @@
+                        {% if tx.description is defined %}
+                            <b>Description:</b> {{tx.description}}<br> 
+                        {% endif %}
+			<b>Label:</b> {{tx.label}}<br>
```
fix spacing

eoliveros replied:
fixed

djpnewton commented on templates/lightning/lightning_transactions.html:
```diff
@@ -0,0 +1,41 @@
+                        {% endif %}
+			<b>Label:</b> {{tx.label}}<br>
+                    {% endif %}
+		    <b>Payment hash:</b> {{tx.payment_hash}}<br>
```
fix spacing

eoliveros replied:
fixed.

djpnewton commented on ln.py:
```diff
@@ -126,12 +124,48 @@ def list_invoices(self):
                 amount_received_msat = invoice["amount_received_msat"]
+                amount_received_sats = _msat_to_sat(amount_received_msat)
                 paid_at = invoice["paid_at"]
+                #paid_date = datetime.datetime.fromtimestamp(paid_at)
```
dead code

eoliveros replied:
removed the dead code

djpnewton commented on ln.py:
```diff
@@ -126,12 +124,48 @@ def list_invoices(self):
+                amount_received_sats = _msat_to_sat(amount_received_msat)
                 paid_at = invoice["paid_at"]
+                #paid_date = datetime.datetime.fromtimestamp(paid_at)
                 paid_date = datetime.datetime.fromtimestamp(paid_at, pytz.timezone('Pacific/Auckland'))
```
remove hard coded timezone

eoliveros replied:
should be changeable via env variable.

djpnewton commented on ln.py:
```diff
@@ -126,12 +124,48 @@ def list_invoices(self):
+            if status == "complete":
+                paid_at = created_at
+                paid_date = datetime.datetime.fromtimestamp(paid_at, pytz.timezone('Pacific/Auckland'))
+                #paid_date = datetime.datetime.fromtimestamp(paid_at)
```
dead code

eoliveros replied:
removed the dead code

djpnewton commented on ln_wallet_endpoint.py:
```diff
@@ -198,14 +197,17 @@ def channel_opener():
-    return render_template('lightning/channel_opener.html')
+    rpc = LnRpc()
+    funds_dict=rpc.list_funds()
+    return render_template('lightning/channel_opener.html', funds_dict=funds_dict)
```
```suggestion
    return render_template('lightning/channel_opener.html', funds_dict=LnRpc().list_funds())
```

eoliveros replied:
removed line 200 and 201. changed line 202 as well.

djpnewton commented on ln_wallet_endpoint.py:
```diff
@@ -155,23 +149,28 @@ def pay_invoice():
+    for unsorted_dict in dict_txs[1]:
+        unsorted_list_dict.append(unsorted_dict)
+    sorted_txs = sorted(unsorted_list_dict, key=lambda d: d["paid_at"], reverse=True)
+    #sorted_txs = sorted(unsorted_list_dict, key=lambda d: d["paid_date"], reverse=True)
```
dead code

eoliveros replied:
deleted dead code.

djpnewton commented:
it will be easier to remove pytz from the code and just use this method (https://confluence.atlassian.com/kb/how-to-set-the-timezone-for-docker-container-976780914.html) that should automatically update all the places we format dates

eoliveros reviewed (commented):
should be resolved.

eoliveros commented:
> it will be easier to remove pytz from the code and just use this method (https://confluence.atlassian.com/kb/how-to-set-the-timezone-for-docker-container-976780914.html) that should automatically update all the places we format dates

not sure how to do that.

djpnewton commented:
> not sure how to do that. The part that sets the date/time is using
> 
> ```
> datetime.datetime.fromtimestamp(timestamp)
> ```
> 
> will default to UTC. thats why using the pytz.timezone is necessary. Unless you are aware of some other way.
> 
> also, theres suddenly a conflict on this push not sure why. This is repo is getting way too big to redo.

https://docs.python.org/3/library/datetime.html#datetime.datetime.fromtimestamp

```If optional argument tz is None or not specified, the timestamp is converted to the platform’s local date and time```

So putting ```TZ=XXX``` in the .env file (see https://confluence.atlassian.com/kb/how-to-set-the-timezone-for-docker-container-976780914.html) should affect every place we do a timestamp conversion using the default tz=None, this is much easier then adding a manual timezone in a hundred different places in the code

eoliveros commented:
> `If optional argument tz is None or not specified, the timestamp is converted to the platform’s local date and time`

```
- added TZ in .env
- added {TZ} in docker-compose.yml
- removed the pytz module in ln.py'
- changed to timestamp({timestamp})
```

TESTED on my local instance and appears to work so i pushed the changes.

djpnewton commented on templates/lightning/broadcast_psbt.html:
```diff
@@ -1,11 +1,11 @@
 {% extends "lightning/layout.html" %}
 {% block title %}Create Invoice{% endblock %}
+{% block navigation %}
```
you should not need this

eoliveros commented on docker-compose.yml:
```diff
@@ -45,6 +45,7 @@ services:
       WINDCAVE_MOCK: ${WINDCAVE_MOCK}
       EXCHANGE_ACCOUNT_MOCK: ${EXCHANGE_ACCOUNT_MOCK}
       REGISTRATION_DISABLE: ${REGISTRATION_DISABLE}
+      TZ: ${TZ}
```
Yes, its for setting up the timezone on the server.

djpnewton commented:
i fixed the merge conflicts
